include/circt/Dialect/Verif/VerifOps.td

```diff
@@ -212,4 +212,85 @@ def YieldOp : VerifOp<"yield", [
   ];
 }
 
+//===----------------------------------------------------------------------===//
+// Verification intent ops
+//===----------------------------------------------------------------------===//
+
+/// BMC -- short for bounded model check.
+def BMC : I32EnumAttrCase<"BMC", 0, "bmc">;
+/// KIP -- short for k-Inductive-Proof.
+def KIP : I32EnumAttrCase<"KIP", 0, "induction">;
+
+def FormalTestTypeAttr : I32EnumAttr<"FormalTestType", "formal test type",
+                                [BMC, KIP]> {
+  let cppNamespace = "circt::verif";
+}
+
+def FormalOp : VerifOp<"formal"> {
+  let summary = "defines a formal verification test";
+  let description = [{
+    This operation defines a formal verification test with potentially additional
+    assertions and assumptions. This should be written by declaring symbolic 
+    values that are then connected to a hardware instance. These symbols can then
+    be reasoned about in additional assertions. The region will then be converted 
+    into btor2, SMTLib, or SystemVerilog and ignored during regular SystemVerilog 
+    emission.   
+      
+    Additional information about how the test should be conducted can be passed 
+    using the two supported attributes. These attributes are `k`, which defines 
+    the bounds of the test, and `type` which defines the type of formal test 
+    being done, i.e. Bounded Model Checking (BMC) or k-Induction.
+  }];
+
+  let arguments = (ins OptionalAttr<APIntAttr>:$k,
+                       OptionalAttr<FormalTestTypeAttr>:$test_type,
+                       OptionalAttr<InnerSymAttr>:$inner_sym);
+
+  let regions = (region SizedRegion<1>:$test_region);
+
+  let assemblyFormat = [{
+    attr-dict  (`sym` $inner_sym^)? `(` (`k` `:` $k^)? (`type` `:` $test_type^)? `)` 
+    $test_region
+  }];
+}
+
+//===----------------------------------------------------------------------===//
+// Formal Verification Ops
+//===----------------------------------------------------------------------===//
+
+def SymbolicValOp : VerifOp<"symbolic_val", [Pure]> {
```

Comment:
Declaring it `pure` can lead to issues when you have multiple `symbolic_val` with the same result type since CSE would be allowed to merge them which would be equivalent to adding a constraint `symbolic_val1 == symbolic_val2`. To fix this, I think an allocation side-effect on the result value should be enough (which would still allow DCE to remove it when the result is never used).

Reply:
Ah right I didn't think about that, thanks for catching it!